### PR TITLE
[tt-train] Use fused SDPA in DeepSeek MLA

### PR DESCRIPTION
**Pull request by mdragulaTT**
## Summary

This PR plugs fused `scaled_dot_product_attention` into TTML DeepSeek MLA, replacing the current composite SDPA path in `MultiHeadLatentAttention`.

The intent is to validate whether the fused SDPA op is ready for the MLA training path, including autograd and `qk_head_dim != v_head_dim`. MLA uses causal attention, so the fused path passes `None` for the mask and lets the kernel generate the causal mask on chip. This avoids materializing/loading the mask and enables the causal balanced-parallelism path when the shape qualifies.

## Why Add A Focused Test?

The full DeepSeek model tests can fail for unrelated MoE/model issues, which makes it hard to tell whether MLA SDPA itself is correct. This PR adds an MLA-only parity test that compares:

- old composite SDPA MLA path
- new fused SDPA MLA path
- forward output
- backward gradients

## Smoke Training Check

Ran 100-step DeepSeek nano Shakespeare training smoke on N300.

Config:

- `tt-train/configs/training_configs/training_shakespeare_nano_deepseek_char.yaml`
- batch sizes `16`, `32`
- sequence lengths `64`, `128`, `256`
- timing excludes first 15 steps
- timing is full training step time: forward + loss + backward + optimizer
- TFLOPS/MFU are the full training-step estimates printed by `train_nanogpt.py`
- fused SDPA uses on-chip causal mask generation, not a materialized mask tensor

| B | S | composite final loss | fused final loss | composite avg step | fused avg step | composite TFLOPS / MFU | fused TFLOPS / MFU | result |
|---:|---:|---:|---:|---:|---:|---:|---:|---|
| 16 | 64 | 2.343750 | 2.359375 | 341.81 ms | 328.15 ms | 0.34 / 0.6% | 0.36 / 0.6% | fused 4.0% faster |
| 16 | 128 | 2.281250 | 2.265625 | 355.41 ms | 342.82 ms | 0.68 / 1.2% | 0.71 / 1.3% | fused 3.5% faster |
| 16 | 256 | 2.312500 | 2.312500 | 385.33 ms | 374.88 ms | 1.34 / 2.4% | 1.38 / 2.5% | fused 2.7% faster |
| 32 | 64 | 2.187500 | 2.171875 | 369.71 ms | 351.41 ms | 0.64 / 1.1% | 0.67 / 1.2% | fused 4.9% faster |
| 32 | 128 | 2.156250 | 2.187500 | 375.13 ms | 373.79 ms | 1.29 / 2.3% | 1.30 / 2.3% | about equal |
| 32 | 256 | 2.109375 | 2.109375 | 545.60 ms | 540.73 ms | 1.89 / 3.4% | 1.91 / 3.4% | about equal |

Loss deltas across the 100 steps were small:

- `B=16, S=64`: max abs delta `0.046875`, mean abs delta `0.011250`
- `B=16, S=128`: max abs delta `0.031250`, mean abs delta `0.010156`
- `B=16, S=256`: max abs delta `0.062500`, mean abs delta `0.009531`
- `B=32, S=64`: max abs delta `0.031250`, mean abs delta `0.009062`
- `B=32, S=128`: max abs delta `0.031250`, mean abs delta `0.007188`
- `B=32, S=256`: max abs delta `0.046875`, mean abs delta `0.008750`

For the attached plots, I also reran the largest smoke shape for 200 steps:

| B | S | steps | composite final loss | fused final loss | composite avg step | fused avg step | composite TFLOPS / MFU | fused TFLOPS / MFU | result |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---|
| 32 | 256 | 200 | 1.601562 | 1.617188 | 553.56 ms | 536.20 ms | 1.87 / 3.3% | 1.92 / 3.4% | fused 3.1% faster |

200-step loss delta: max abs `0.046876`, mean abs `0.012734`, final delta `+0.015626`.

### Plot: Loss Curve

loss_curve_200_b32_s256_causal

### Plot: Loss Delta

loss_delta_200_b32_s256_causal

### Plot: Step Time

step_time_200_b32_s256_causal

## Notes

Correctness looks reasonable from the focused parity test and training smoke. Performance is non-regressing in this full-step smoke: fused SDPA is faster for the smaller cases and about equal for `B=32, S=128/256`.

I am leaving this for review/decision because the remaining question is whether reviewers want additional isolated MLA/SDPA forward/backward timing before merging the fused path.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/mla_sdpa_plug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/mla_sdpa_plug)

### CODE REVIEW

File: tt-train/tests/python/test_deepseek_mla_sdpa.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    if metrics["abs_mean"] >= 1e-5:
+        assert metrics["cos_sim"] > 0.995, f"{name} cosine similarity too low: {metrics}"
+
+
+@pytest.fixture(autouse=True)
+def reset_graph():
+    yield
+    ttml.autograd.AutoContext.get_instance().reset_graph()
+
+
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/tests/python/test_deepseek_mla_sdpa.py:136]: Ok. Added.

File: tt-train/sources/ttml/ttml/models/deepseek/mla.py [hunk 2]
```diff
@@ -113,8 +113,10 @@ def forward(self, x: ttml.autograd.Tensor, mask: ttml.autograd.Tensor) -> ttml.a
         q_full = autograd_concat([q_nope, q_pe], dim=3)  # [B, H, S, qk_head]
         k_full = autograd_concat([k_nope, k_pe], dim=3)  # [B, H, S, qk_head]
 
-        # ── Attention (composite path supports v_dim != qk_head) ──
-        attn = ttml.ops.attention.scaled_dot_product_attention_composite(q_full, k_full, v, mask)
+        # ── Attention ──
+        # MLA uses a causal mask; passing None lets the fused SDPA kernels generate it on chip
+        # and select their causal/balanced path instead of loading a materialized arbitrary mask.
+        attn = ttml.ops.attention.scaled_dot_product_attention(q_full, k_full, v, None)
```
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/mla.py]: Not related to this PR, but surprised `None` is casual mask in our API.
In Pytorch default is full attention.

More importantly `MultiHeadLatentAttention.forward` this function takes mask as an argument. We silently reject it and hardcode `None`.

We need to either pass the mask argument to SDPA or drop the argument and document this as casual only.

File: tt-train/sources/ttml/ttml/models/deepseek/mla.py
Comment (zbaczewskiTT) [inline tt-train/sources/ttml/ttml/models/deepseek/mla.py]: Not updated; now it's fused and mask is ignored.

File: tt-train/tests/python/test_deepseek_mla_sdpa.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        ctx.set_gradient_mode(ttml.autograd.GradMode.DISABLED)
+        x, mask = _make_inputs()
+        out_composite = _old_composite_forward(module, x, mask)
+        ttnn.synchronize_device(device)
+        out_composite_np = _to_numpy(out_composite)
+        ctx.reset_graph()
+
+        x, mask = _make_inputs()
+        out_fused = module(x, mask)
```
Comment (zbaczewskiTT) [inline tt-train/tests/python/test_deepseek_mla_sdpa.py]: We create a mask and pass it into MLA, but it's always ignored. This is confusing.

**Review by zbaczewskiTT — CHANGES_REQUESTED** [review]
Thank you!
Please correct the noted stuff.

File: tt-train/sources/ttml/ttml/models/deepseek/mla.py
Comment (mdragulaTT) [reply tt-train/sources/ttml/ttml/models/deepseek/mla.py]: Done.

File: tt-train/tests/python/test_deepseek_mla_sdpa.py (hunk 3 above)
Comment (mdragulaTT) [reply tt-train/tests/python/test_deepseek_mla_sdpa.py]: Done.

File: tt-train/sources/ttml/ttml/models/deepseek/mla.py (hunk 2 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/ttml/models/deepseek/mla.py]: Addressed — `MultiHeadLatentAttention.forward` no longer takes `mask` (it was the silent-ignore you flagged); MLA is now explicitly causal-only and generates the causal mask on chip.

I kept mask on the block and `DeepSeek.forward` deliberately — it's pinned by two framework contracts: the shared `forward(input, mask)` model interface the trainer calls polymorphically (`train_nanogpt.py`), and `memory_efficient_runner`, which always invokes `block(input, mask, …)`. Removing it there would mean special-casing DeepSeek in shared training code, so I documented it as unused/causal-only instead.

On "do we even need a mask": for DeepSeek as trained here, no — it's causal-only. A real (non-causal) mask would only matter for sequence packing (document/block-diagonal masks) or padding (variable-length batches), neither of which the tt-train DeepSeek path uses today. So I'd rather not rip it out now; tracked the broader "should DeepSeek expose a mask at all" as a future cleanup.

Comment (zbaczewskiTT) [reply tt-train/sources/ttml/ttml/models/deepseek/mla.py]: Thank you

**Review by zbaczewskiTT — APPROVED** [review]

File: tt-train/tests/python/test_deepseek_mla_sdpa.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    """Create deterministic MLA activations plus the explicit mask used by the old composite path."""
+    rng = np.random.default_rng(SEED)
+    x_np = rng.standard_normal((batch_size, 1, seq_len, dim), dtype=np.float32)
+    mask_np = np.tril(np.ones((seq_len, seq_len), dtype=np.float32)).reshape(1, 1, seq_len, seq_len)
+    x = ttml.autograd.Tensor.from_numpy(x_np, layout=ttnn.Layout.TILE, new_type=ttnn.DataType.BFLOAT16)
+    mask = ttml.autograd.Tensor.from_numpy(mask_np, layout=ttnn.Layout.TILE, new_type=ttnn.DataType.BFLOAT16)
+    return x, mask
+
+
+def _old_composite_forward(module: MultiHeadLatentAttention, x: ttml.autograd.Tensor, mask: ttml.autograd.Tensor):
```
Comment (philei-tt) [inline tt-train/tests/python/test_deepseek_mla_sdpa.py:56]: Do we really need this test? I think sdpa tests should already be covering fused vs composite comparison.

**Review by philei-tt — APPROVED** [review]

File: tt-train/tests/python/test_deepseek_mla_sdpa.py (hunk 4 above)
Comment (mdragulaTT) [reply tt-train/tests/python/test_deepseek_mla_sdpa.py:56]: Agree, but this isn’t meant to test SDPA correctness. It’s an MLA integration test: checks that MLA calls fused SDPA with the right shape (`qk_head_dim != v_head_dim`), causal path (`mask=None`), and matches the old composite MLA path. I'd keep it.
